src/streaming/array.writer.cpp

```diff
@@ -0,0 +1,422 @@
+#include "macros.hh"
+#include "array.writer.hh"
+#include "zarr.common.hh"
+#include "zarr.stream.hh"
+#include "sink.creator.hh"
+
+#include <cmath>
+#include <functional>
+#include <latch>
+#include <stdexcept>
+
+#ifdef min
+#undef min
+#endif
+
+bool
+zarr::downsample(const ArrayWriterConfig& config,
+                 ArrayWriterConfig& downsampled_config)
+{
+    // downsample dimensions
+    std::vector<ZarrDimension> downsampled_dims;
+    for (auto i = 0; i < config.dimensions->ndims(); ++i) {
+        const auto& dim = config.dimensions->at(i);
+        // don't downsample channels
+        if (dim.type == ZarrDimensionType_Channel) {
+            downsampled_dims.push_back(dim);
+        } else {
+            const uint32_t array_size_px =
+              (dim.array_size_px + (dim.array_size_px % 2)) / 2;
+
+            const uint32_t chunk_size_px =
+              dim.array_size_px == 0
+                ? dim.chunk_size_px
+                : std::min(dim.chunk_size_px, array_size_px);
+
+            CHECK(chunk_size_px);
+            const uint32_t n_chunks =
+              (array_size_px + chunk_size_px - 1) / chunk_size_px;
+
+            const uint32_t shard_size_chunks =
+              dim.array_size_px == 0
+                ? 1
+                : std::min(n_chunks, dim.shard_size_chunks);
+
+            downsampled_dims.emplace_back(dim.name,
+                                          dim.type,
+                                          array_size_px,
+                                          chunk_size_px,
+                                          shard_size_chunks);
+        }
+    }
+    downsampled_config.dimensions = std::make_shared<ArrayDimensions>(
+      std::move(downsampled_dims), config.dtype);
```

Comment:
Why are you using a shared pointer here? A unique pointer should suffice since `downsampled_config` is its exclusive owner.